New Serverless Pattern - Selective file transfer between SFTP server and S3 using AWS Transfer Family (#2267)
aws-samples/serverless-patterns: awstransfer-s3-sam/README.md

# Bidirectional selective file transfer between remote SFTP server and Amazon S3 using AWS Transfer Family Connector

> This pattern shows how to setup an AWS Transfer Family SFTP connector to list files from the remote server and transfer specific file to Amazon S3 bucket. You can also transfer specific file from Amazon S3 bucket to the remote SFTP Server.  

Reviewer: ```suggestion
This pattern shows how to setup an AWS Transfer Family SFTP connector to list files from the remote server and transfer specific files to Amazon S3 bucket. You can also transfer specific files from Amazon S3 bucket to the remote SFTP server.  
```

Author: Fixed

* [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have sufficient permissions to make necessary AWS service calls and manage AWS resources.
* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
* [AWS Serverless Application Model](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) (AWS SAM) installed
* [JQ](https://docs.aws.amazon.com/solutions/latest/dynamic-object-and-rule-extensions-for-aws-network-firewall/operation-and-customization.html#install-jq) should be installed.

## Deployment Instructions

1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
    ``` 
    git clone https://github.com/aws-samples/serverless-patterns
    ```

2. Change directory to the pattern directory:
    ```
    cd serverless-patterns/awstransfer-s3-sam
    ```

3. From the command line, run the below command to deploy the pattern:
    ```

>     bash deploy.sh

Reviewer: How would a customer using Windows do this?

Author: Customer using Windows system can use any command line tool that support using using bash script. I used to use GitBash but there may other alternative options also:
https://stackoverflow.com/questions/26522789/how-to-run-sh-on-windows-command-prompt

    ```

4. During the prompts:
    * Enter a stack name

>     * Enter the desired AWS Region

Reviewer: Please add in which format the user should enter it

5. The deployment script deploys both `template1.yaml` and `template2.yaml`. Please make a note of the output both the deployments as they will be used during testing.

## How it works

Please refer to the architecture diagram below:

![End to End Architecture](images/architecture.png)

* The remote SFTP server is simulated using AWS Transfer Family SFTP Server for this pattern. In real use case, this can be any remove SFTP server outside of AWS. 
* SFTP Connector is configured to comment the remote server with Amazon S3 bucket using SFTP protocol. The authentication is done using SSH Key based handshake. 

> * Amazon S3 bucket file storage on AWS side. 

Reviewer: ```suggestion
* Amazon S3 bucket for storage on AWS side. 
```

Author: Fixed.

> * User can list files on the remote server and selectively transfer file from the remote server to the Amazon S3 bucket using AWS Transfer Family API or CLI commands. 

Reviewer: ```suggestion
* User can list files on the remote server and selectively transfer files from the remote server to the Amazon S3 bucket using AWS Transfer Family API or CLI commands. 
```

Author: Fixed.

    ```bash
    aws transfer test-connection --region {your-region} --connector-id {SFTPTransferConnector}
    ```
    It should give an output similar to below:
    ```json
    {
        "Status": "OK",
        "StatusMessage": "Connection succeeded"
    }
    ```

3. Transfer `sample1.txt` and `sample2.txt` files to the remote SFTP server (similated) using the below commands. In this sample project replace `SSHPrivateKeyFileName` and `SSHPrivateKeyFileName` with `sftpuser`. Replace the value of `TransferServerEndpoint` from the deployment output:
    ```bash
    sftp -i {SSHPrivateKeyFileName} {TransferServerUser}@{TransferServerEndpoint}
    pwd
    mkdir Remote
    cd Remote
    put sample1.txt
    put sample2.txt
    ls
    ```
    Confirm to proceed with the connection after the first command.

5. List the files on the remote SFTP server using the below command. Please replace `SFTPTransferConnector` and `MyLocalS3Bucket` from the deploy output: 

    ```bash
    aws transfer start-directory-listing --region {your-region} --connector-id {SFTPTransferConnector} --remote-directory-path /Remote --output-directory-path /{MyLocalS3Bucket}/FromRemoteSFTPServer
    ```
    

>     The command invokes async API. The outpul of the command will be as follows:

Reviewer: ```suggestion
    The command invokes an asynchronous API. The output of the command will be as follows:
```

Author: Fixed

    ```json
    {
    "ListingId": "273e5b33-xxxx-xxxx-xxxx-xxxxx9a507f53",
    "OutputFileName": "c-cxxxxxxxx-xxxxxx-xxxx-xxxx-xxxx-xxxxxa507f53.json"
    }
    ```

6. Log into [Amazon S3 console](https://console.aws.amazon.com/s3). Open the `MyLocalS3Bucket` and navigate to `FromRemoteSFTPServer` folder. Check the content of the JSON file. It should look something like below:
    ```json
    {
    "files": [
        {
        "filePath": "/Remote/sample1.txt",
        "modifiedTimestamp": "2024-04-28T07:47:27Z",
        "size": 146
        },
        {
        "filePath": "/Remote/sample2.txt",
        "modifiedTimestamp": "2024-04-28T07:47:46Z",
        "size": 146
        }
    ],
    "paths": [],
    "truncated": false
    }
    ```

> 7. Transfer one of the files from the remote SFTP server to Amazon S3 bucket using the following command:

Reviewer: ```suggestion
7. Transfer one of the files from the remote SFTP server to the Amazon S3 bucket using the following command:
```

Author: Fixed.
